packages/@aws-cdk/aws-servicecatalog/lib/template.ts

```diff
@@ -0,0 +1,93 @@
+import * as s3_assets from '@aws-cdk/aws-s3-assets';
+import * as cdk from '@aws-cdk/core';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * Represents the Product Provisioning Artifact Template.
+ */
+export abstract class Template {
+  /**
+   * Template from URL
+   * @param url The url that points to the provisioning artifacts template
+   */
+  public static fromUrl(url: string): UrlTemplate {
+    return new UrlTemplate(url);
+  }
+
+  /**
+   * Loads the provisioning artifacts template from a local disk path.
+   *
+   * @param path A file containing the provisioning artifacts
+   */
+  public static fromAsset(path: string, options?: s3_assets.AssetOptions): AssetTemplate {
+    return new AssetTemplate(path, options);
+  }
+
+  /**
+   * Called when the product is initialized to allow this object to bind
+   * to the stack, add resources and have fun.
+   *
+   * @param scope The binding scope. Don't be smart about trying to down-cast or
+   * assume it's initialized. You may just use it as a construct scope.
+   */
+  public abstract bind(scope: Construct): TemplateConfig;
+}
+
+/**
+ * Result of binding `Template` into a `Product`.
+ */
+export interface TemplateConfig {
+  /**
+    * The http url of the template in S3.
+    */
+  readonly httpUrl: string;
+}
+
+/**
+ * Template code from a Url.
+ */
+export class UrlTemplate extends Template {
+  constructor(private url: string) {
+    super();
+  }
+
+  public bind(_scope: Construct): TemplateConfig {
+    return {
+      httpUrl: this.url,
+    };
+  }
+}
+
+/**
+ * Template from a local file.
+ */
+export class AssetTemplate extends Template {
+  private asset?: s3_assets.Asset;
+
+  /**
+   * @param path The path to the asset file.
+   */
+  constructor(public readonly path: string, private readonly options: s3_assets.AssetOptions = { }) {
+    super();
+  }
+
+  public bind(scope: Construct): TemplateConfig {
+    // If the same AssetCode is used multiple times, retain only the first instantiation.
+    if (!this.asset) {
+      this.asset = new s3_assets.Asset(scope, 'Template', {
+        path: this.path,
+        ...this.options,
+      });
+    } else if (cdk.Stack.of(this.asset) !== cdk.Stack.of(scope)) {
+      throw new Error(`Template is already associated with another stack '${cdk.Stack.of(this.asset).stackName}'. ` +
```

Comment:
Why is this validation here? What's wrong with using the same `Template` for multiple Products in different Stacks?

Reply:
This validation was taken from Lambda Code. 
I was wondering if there was an underlying reason why Lambda Assets could only be associated with one stack and if it applied to our use case as well?
Since users can easily create another instance for another stack the additional validation doesn't affect functionality, so I added it to be safe, but if you are sure it is not needed I can remove it.

Reply:
Yeah, let's ditch it for now. I don't see a problem with using the same `Template` multiple times (a unit test exercising that scenario would be nice!).

Reply:
Oh, I think I understand now. This is for cases like:

```ts
const myTemplate = 
new CloudFormationProduct(this, 'P1', {
  provisioningArtifacts: [
    { template: myTemplate }
  ],
});
new CloudFormationProduct(this, 'P2', {
  provisioningArtifacts: [
    { template: myTemplate }
  ],
});
```

In this case, this class is trying to only have a single Asset created in the template.

My problem with that is that it will fail for the almost equivalent:

```ts
new CloudFormationProduct(this, 'P1', {
  provisioningArtifacts: [
    { template: Template.fromAsset('template.json') }
  ],
});
new CloudFormationProduct(this, 'P2', {
  provisioningArtifacts: [
    { template: Template.fromAsset('template.json') }
  ],
});
```

Which means I don't think it's a particularly robust validation.

I'm fine with punting on theses edge-cases in this PR. The "proper" solution will probably require more work.

Reply:
I thought we agreed to remove the validation for now...?

Reply:
Sorry I misunderstood, I'll remove it